Cleanup KK warning system

jgfouca opened the pull request:
Change list:
1) Bring over similar warning system from main kokkos, except encapsulate it into a macro `kokkoskernels_enable_warnings` which was added to new file `kokkoskernels_util.cmake`
2) Formatting improvements, especially in regards to remove all-caps, in a couple cmake files
3) Fix warnings all over the place
4) Add new KokkosKernels options `KokkosKernels_ENABLE_COMPILER_WARNINGS` and  `KokkosKernels_ENABLE_WARNINGS_AS_ERRORS`.
5) Change CI to use these options instead of hardcoding cxx flags.

jgfouca committed:
Cleanup KK warning system

Signed-off-by: James Foucar <jgfouca@sandia.gov>

jgfouca commented on CMakeLists.txt:
```diff
@@ -1,170 +1,190 @@
-        OFF
-        BOOL
-        "Whether to build docs. Default: OFF"
+kokkoskernels_add_option(
```
Real new code is here.

jgfouca commented on batched/dense/src/KokkosBatched_Vector_SIMD.hpp:
```diff
@@ -76,7 +76,20 @@ class Vector<SIMD<T>, l> {
   KOKKOS_INLINE_FUNCTION
   value_type &operator[](const int &i) const { return _data[i]; }
+
+  KOKKOS_INLINE_FUNCTION
```
These changes are worth a close review as they potentially change code behavior.

jgfouca committed:
Nix another warning

Signed-off-by: James Foucar <jgfouca@sandia.gov>

lucbv commented:
I have some general comments:

- If you want this do be a reasonably fast PR I would only add the current set of warnings listed in our osx build: `-Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wuninitialized`. Other subsequent PRs can propose to add more warnings later on
- The switch from upper to lower case in CMakeLists.txt is arguably an improvement but it also makes it fairly hard to know what changed in this PR, I would suggest to make that change separately, maybe after consulting everyone in the weekly meeting?
- There are still some issues in the OSX build with the new warning and there are new warnings emitted in the CUDA/HIP builds but we are not using `-Werror` there?

jgfouca commented:
@lucbv ,

> If you want this do be a reasonably fast PR I would only add the current set of warnings listed in our osx build: -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wuninitialized. Other subsequent PRs can propose to add more warnings later on

I'm in no hurry, so we can take the time to do this the way we want it long term. I went through and fixed the new warnings that were discovered in our code. Most of them were extra semicolons which were easy to fix. The only tricky one was "deprecated copy", which required some changes to our Vector class. I think there's probably some value in having the same warnings on as Kokkos core since we are in that ecosystem.

> The switch from upper to lower case in CMakeLists.txt is arguably an improvement but it also makes it fairly hard to know what changed in this PR, I would suggest to make that change separately, maybe after consulting everyone in the weekly meeting?

I agree. I tried to flag the "real" change via gh comment. The only real change to the main CMakeLists.txt file is this:
```
kokkoskernels_add_option(
  "ENABLE_COMPILER_WARNINGS"
  OFF
  BOOL
  "Whether to turn on more warnings. Default: OFF"
)

kokkoskernels_add_option(
  "ENABLE_WARNINGS_AS_ERRORS"
  OFF
  BOOL
  "Whether warnings should be treated as errors. Default: OFF"
)

include(cmake/kokkoskernels_util.cmake)

... Later, in the KK standalone build section:

    if (KokkosKernels_ENABLE_COMPILER_WARNINGS)
      kokkoskernels_enable_warnings()
    endif()
```

I also added a macro for turning on warnings (code that was copied with minor mods from Kokkos) in the new file `kokkoskernels_util.cmake`.

> There are still some issues in the OSX build with the new warning and there are new warnings emitted in the CUDA/HIP builds but we are not using -Werror there?

I will clean these issues up.

jgfouca committed:
Fix another trailing semi

Signed-off-by: James Foucar <jgfouca@sandia.gov>

jgfouca committed:
Fix a few mistakes

1) Warnings should only be set up for KK standalone and
   AFTER the Kokkos package is loaded.
2) Add NVCC check to guard against undefined pragmas in KokkosBlas1_reciprocal_impl.hpp.
   Also, encapsulate the need for de-optimization in one variable.
3) Fix casing of kokkos compiler ID checks

Signed-off-by: James Foucar <jgfouca@sandia.gov>

jgfouca commented on blas/impl/KokkosBlas1_reciprocal_impl.hpp:
```diff
@@ -14,6 +14,12 @@ namespace Impl {
 // reciprocal
 //
 
+#if defined(__GNUC__) && ((__GNUC__ == 12) || (__GNUC__ == 13)) && !defined(__NVCC__)
```
This is a change worth reviewing as well.

ndellingwood commented on blas/impl/KokkosBlas1_reciprocal_impl.hpp:
```diff
@@ -14,6 +14,12 @@ namespace Impl {
 //
 
+#if defined(__GNUC__) && ((__GNUC__ == 12) || (__GNUC__ == 13)) && !defined(__NVCC__)
+#define NEED_TO_DEOPT 1
```
might need to update the macro naming here, maybe something like `KOKKOSKERNELS_GCC_BUG_DISABLE_TREE_VECTORIZE` to add more detail?

jgfouca replied:
Done.

jgfouca committed:
Better name for cpp var

Signed-off-by: James Foucar <jgfouca@sandia.gov>

lucbv commented on batched/dense/src/KokkosBatched_Vector_SIMD.hpp:
```diff
@@ -76,7 +76,20 @@ class Vector<SIMD<T>, l> {
   value_type &operator[](const int &i) const { return _data[i]; }
+
+  KOKKOS_INLINE_FUNCTION
+  type &operator=(const type &rhs) {
```
Why not allow the same type of input as the copy constructor?

```
template <typename ArgValueType>
  KOKKOS_INLINE_FUNCTION
    type &operator=(const Vector<SIMD<ArgValueType>, vector_length> &rhs) {
    KOKKOSKERNELS_FORCE_SIMD
    for (int i = 0; i < vector_length; ++i) _data[i] = rhs._data[i];
  }
```

lucbv commented on batched/dense/src/KokkosBatched_Vector_SIMD.hpp:
```diff
@@ -76,7 +76,20 @@ class Vector<SIMD<T>, l> {
+    for (int i = 0; i < vector_length; ++i) _data[i] = b[i];
+  }
 };
+
```
Don't you need a destructor for `Vector`?

lucbv commented on cmake/kokkoskernels_warnings.cmake:
I do not like the name of this file, `util` is the least descriptive name and will lead to others lazily adding anything to it until we have no clue what this file is intended to provide.

lucbv commented on CMakeLists.txt:
```diff
@@ -1,170 +1,190 @@
+      message(FATAL_ERROR "Kokkos Kernels ${KokkosKernels_VERSION} requires Kokkos 4.6.02 or greater (found ${Kokkos_VERSION})")
+    endif()
+
+    if (KokkosKernels_ENABLE_COMPILER_WARNINGS)
```
I might be wrong but aren't you supposed to write cmake logic using the cache variable instead of the user passed parameter?

```suggestion
    if (KOKKOSKERNELS_ENABLE_COMPILER_WARNINGS)
```

Note that I could be wrong here, so please do not just make the change, it would be good to understand what is the proper way to do this

jgfouca replied:
`KokkosKernels_XXX` is the CACHE variable:

```
  SET(CAMEL_NAME KokkosKernels_${SUFFIX})
  STRING(TOUPPER ${CAMEL_NAME} UC_NAME)

  # Make sure this appears in the cache with the appropriate DOCSTRING                                                                                                                                                          
  SET(${CAMEL_NAME} ${DEFAULT} CACHE ${TYPE} ${DOCSTRING})
```

There's a lot of stuff in `kokkoskernels_add_option` that doesn't make sense to me and is not standard CMake. I assume this has something to do with tribits. It looks like we make sure that the cache is always camel case but then set non-cache variables in all caps. I assume this is to allow case-insensitive cmake code. In other words, both

`if (KokkosKernels_XXX)` and `if (KOKKOSKERNELS_XXX)` will work. I think it's better to just use the CACHE variable (which is also what the user would pass).

lucbv reviewed (changes requested):
Some questions and some changes required but nothing major

jgfouca committed:
Address GH feedback. Rename cmake include and some minor fixes to Vector

Signed-off-by: James Foucar <jgfouca@sandia.gov>

jgfouca committed:
Formatting

Signed-off-by: James Foucar <jgfouca@sandia.gov>

lucbv commented:
@jgfouca the last PR I merged created a small conflict, can you have a look, I'm happy to review again after

jgfouca committed:
Merge remote-tracking branch 'upstream/develop' into HEAD

* upstream/develop:
  CI: building against 5.0.0 in CI and update our version (#2862)
  Add missing popRegion() (#2863)
  More config options for par_ilut benchmark (#2859)

Signed-off-by: James Foucar <jgfouca@sandia.gov>

jgfouca commented:
@lucbv , done
